File: Medium/Coin Change/coin-change.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;
class GfG {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            int sum = sc.nextInt();
            int N = sc.nextInt();
            int coins[] = new int[N];
            for (int i = 0; i < N; i++) coins[i] = sc.nextInt();
            Solution ob = new Solution();
            System.out.println(ob.count(coins, N, sum));
        }
    }
}


// } Driver Code Ends


// User function Template for Java

class Solution {
      public long count(int coins[], int n, int sum) {
        // code here.
        long [][]dp=new long [n+1][sum+1];
        Arrays.fill(dp[0],0);
        int i,j;
        for(i=1;i<=n;i++)
        {
            dp[i][0]=1;
        }
        for(i=1;i<=n;i++)
        {
            for(j=1;j<=sum;j++)
            {
                if(j-coins[i-1]>=0)
                {
                    dp[i][j]=dp[i-1][j]+dp[i][j-coins[i-1]];
                }
                else
                {
                    dp[i][j]=dp[i-1][j];
                }
            }
        }
        return dp[n][sum];
    }
}

File: Medium/Coin Change/coin-change.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends



class Solution {
public:
    long long int helper(int idx, int coins[], int sum, long long int dp[1001][1001]) {
        if (idx == 0) {
            if (sum % coins[idx] == 0)
                return dp[idx][sum] = 1LL;
            return dp[idx][sum] = 0LL;
        }
        if (dp[idx][sum] != -1)
            return dp[idx][sum];
        long long int take = 0;
        long long int notTake = helper(idx - 1, coins, sum, dp);
        if (sum >= coins[idx])
            take = helper(idx, coins, sum - coins[idx], dp);
        return dp[idx][sum] = take + notTake;
    }
    
    long long int count(int coins[], int n, int sum) {
        long long int dp[1001][1001];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= sum; j++) {
                dp[i][j] = -1;
            }
        }
        return helper(n - 1, coins, sum, dp);
    }
};





//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int sum, N;
        cin >> sum >> N;
        int coins[N];
        for (int i = 0; i < N; i++) cin >> coins[i];
        Solution ob;
        cout << ob.count(coins, N, sum) << endl;
    }

    return 0;
}


// } Driver Code Ends